### PR TITLE
Port editor files to webpack (no more bower!)

### PR DESCRIPTION
**Pull request by pamelafox**
### High-level description of change

The previous PR ported the output side to webpack, this PR ports the editor side. This PR removes bower entirely and removes build-related tasks from gulp (it leaves test tasks there). 

This branch seems to fully work - all tests are run and are passing, and the demos function. It may be possible to get it working in webapp before moving to React. I'll give that a shot after this PR is merged into my branch.

Notable changes in this PR, besides change to import/export:
* ACE upgraded: there was a change in the format of their deltas, so that required changing some code, mostly in editor-ace and in some tooltips.
* Tomfoolery with images: Getting images to load from url()s in CSS is not so easy. It seems to be working now, thanks to some loaders, but it did involve some URL changes. I imagine that may need tweaking when I build for webapp.
* Module split: I broke TooltipBase into its own file, as it was sharing a file with TooltipEngine. That looks like a large change, but it's really just a file split. I left ImageModal and SoundModal in the same file, as that would have looked even bigger (and there are some actual code changes within them). I'm happy to merge/split as you'd prefer for this diff.
* Tooltip Registration: I added a registerTooltip function to register each tooltip class with the TooltipEngine, as that seems to be a bit better than the old technique (of adding them to TooltipEngine.classes directly).
* Lodash to Underscore: I realized that webapp uses underscore, and there are tiny API differences that matter, so I changed all the imports to underscore, across the editor side and output side.

### Are there performance implications for this change?

This is likely making a bigger bundle locally, as there's a lot that goes into the editor- jquery, backbone, bootstrap, jquery plugins, underscore. 
In webapp, the bundle wouldn't be that large (due to more externals) though I don't know how it'd be post the webpack change.
The react change removes all those dependencies except underscore.

### Have you added tests to cover this new/updated code?

No new tests, got old tests running.

### Risks involved

Big change, easy to break things. I'm most wary of CSS/image/plugins related issues.

### Are there any dependencies or blockers for merging this?

No blockers for merging into my branch. Can't be merged to master, however, as this isn't necessarily webapp compatible.

### How can we verify that this change works?

Run the demos, run the tests.

### CODE REVIEW

File: js/shared/i18n.js [hunk 1]
```diff
@@ -82,7 +82,7 @@ import Jed from "jed";
         // Replace the substitutions with the appropriate option
         for (var i = 1; i < split.length; i += 2) {
             var replaceWith = options[split[i]];
-            split[i] = _.isUndefined(replaceWith) ?
+            split[i] = (replaceWith === undefined) ?
```
Comment (pamelafox) [inline js/shared/i18n.js]: This is what i18n has in webapp, seemed better to change it than bring in underscore.

File: js/ui/tooltip-base.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
+import $ from "jquery";
+import Backbone from "backbone";
+Backbone.$ = $;
+
+import "../../css/ui/tooltips.css";
+
```
Comment (pamelafox) [inline js/ui/tooltip-base.js]: The file split from tooltip-engine.

File: js/ui/tooltips/auto-suggest.js [hunk 3]
```diff
@@ -43,7 +52,6 @@ TooltipEngine.classes.autoSuggest = TooltipBase.extend({
             this.updateTooltip(lookupParams);
             this.placeOnScreen();
             event.stopPropagation();
-            ScratchpadAutosuggest.enableLiveCompletion(false);
```
Comment (pamelafox) [inline js/ui/tooltips/auto-suggest.js]: LiveCompletion is always off now, so I could remove these.

File: js/ui/tooltips/color-picker.js [hunk 4]
```diff
@@ -108,8 +118,8 @@ TooltipEngine.classes.colorPicker = TooltipBase.extend({
             addSemicolon = true;
         }
 
-        if (event.source && event.source.action === "insertText" &&
-            event.source.text.length === 1 && this.parent.options.type === "ace_pjs" && this.autofill) {
+        if (event.source && event.source.action === "insert" &&
```
Comment (pamelafox) [inline js/ui/tooltips/color-picker.js]: ACE upgrade related

File: js/ui/tooltips/image-picker.js [hunk 5]
```diff
@@ -32,7 +39,7 @@ TooltipEngine.classes.imagePicker = TooltipBase.extend({
 
         // TODO(kevinb) extract this into a method on TooltipBase
         if (leadingPadding.length === 0 && path.length === 0 && this.closing.length === 0 &&
-            event.source && event.source.action === "insertText" && event.source.text.length === 1 && this.autofill) {
+            event.source && event.source.action === "insert" && event.source.lines[0].length === 1 && this.autofill) {
```
Comment (pamelafox) [inline js/ui/tooltips/image-picker.js]: ACE upgrade related

File: js/ui/tooltips/number-scrubber.js [hunk 6]
```diff
@@ -1,5 +1,10 @@
-// A description of general tooltip flow can be found in tooltip-engine.js
-TooltipEngine.classes.numberScrubber = TooltipBase.extend({
+import $ from "jquery";
+import "webpack-jquery-ui/draggable";
```
Comment (pamelafox) [inline js/ui/tooltips/number-scrubber.js]: I found a library that claimed to make it easy to bring in jquery-ui. It seems to work.

File: tests/tooltips/colorPicker_test.js [hunk 7]
```diff
@@ -1,55 +1,81 @@
+import {getLine, getMockedTooltip, getTooltip, testMockedTooltipDetection, testReplace, typeLine} from "./shared.js";
+import ColorPicker from "../../js/ui/tooltips/color-picker.js";
+
 describe("colorPicker - detection", function() {
-    var mockedColorPicker = getMockedTooltip(tooltipClasses.colorPicker, ["detector", "initialize"]);
+
+    let sandbox;
```
Comment (pamelafox) [inline tests/tooltips/colorPicker_test.js]: I changed the tooltip tests to use sinon sandbox. It seemed safer now that tests are being bundled together, and its generally a good practice so that you can restore.

File: tmpl/live-editor.handlebars [hunk 8]
```diff
@@ -17,7 +17,7 @@
 
             <div class="scratchpad-canvas-loading">
                 <img src="{{imagesDir}}/spinner-large.gif">
-                <span class="hide-text">{{#_}}Loading...{{/_}}</span>
+                <span class="hide-text">Loading...</span>
```
Comment (pamelafox) [inline tmpl/live-editor.handlebars]: I removed the {{#_}} calls since they don't actually get handled by webapp (as you recall from my PRs a few weeks ago, and this template in particular gets completely overriden by webapp anyway.

File: css/ui/tooltips.css [hunk 9]
```diff
@@ -288,11 +288,11 @@ html.playing .tooltip.hide-while-playing {
 
 /** Media picker CSS: Used for both an image and audio picker **/
 .mediapicker-modal {
+    background-color: #fff;
```
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply css/ui/tooltips.css]: Eventually this will be replaced with wonder-blocks-modal, right?

File: js/shared/i18n.js (hunk 1 above)
Comment (kevinbarabash) [reply js/shared/i18n.js]: sgtm... we should really create our own i18n npm package so that it's easy to use here and elsewhere.

File: js/ui/tooltips/auto-suggest.js (hunk 3 above)
Comment (kevinbarabash) [reply js/ui/tooltips/auto-suggest.js]: Why did we turn off auto-complete?

File: .babelrc [hunk 10]
```diff
@@ -0,0 +1,4 @@
+{
+  "presets": ["env"],
```
Comment (kevinbarabash) [inline .babelrc]: I thought `babel-preset-env` required a `.browserslistrc` to work.  Does it have some sort of default?

File: .travis.yml [hunk 11]
```diff
@@ -3,8 +3,7 @@ sudo: false
 node_js:
   - "8.11.3"
 before_script:
-  - npm install -g bower
-  - bower install
+  - npm install
```
Comment (kevinbarabash) [inline .travis.yml]: yay

File: css/ui/colorpicker.css [hunk 12]
```diff
@@ -3,7 +3,7 @@
 	height: 176px;
 	overflow: hidden;
 	position: absolute;
-	background: url(../images/scratchpads/colorpicker_background.png);
+	background: url(../../images/scratchpads/colorpicker_background.png);
```
Comment (kevinbarabash) [inline css/ui/colorpicker.css]: Why the change in path?

File: demos/simple/audio.html [hunk 13]
```diff
@@ -2,10 +2,6 @@
 <head>
     <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
     <title>Live Editor Audio Demo</title>
-    <link rel="stylesheet" href="../../build/css/live-editor.core_deps.css"/>
-    <link rel="stylesheet" href="../../build/css/live-editor.audio.css"/>
-    <link rel="stylesheet" href="../../build/css/live-editor.tooltips.css"/>
-    <link rel="stylesheet" href="../../build/css/live-editor.ui.css"/>
```
Comment (kevinbarabash) [inline demos/simple/audio.html]: Why are these .css files no longer necessary?

File: js/shared/i18n.js (hunk 1 above)
Comment (kevinbarabash) [reply js/shared/i18n.js]: it looks like this file was just assuming that `_` was a global.  I'm glad we're moving past those dark days.

File: js/ui/autosuggest-data.js [hunk 14]
```diff
@@ -906,3 +908,5 @@ window.ScratchpadAutosuggestData = {
             "keyPressed", "keyReleased", "keyTyped"]
     },
 };
+
+export default AutosuggestData;
```
Comment (kevinbarabash) [inline js/ui/autosuggest-data.js]: You mentioned that we had turned of auto-suggest... do we still need to the files or are you planning on re-enabling and/or making it configurable in the future?

File: js/ui/tooltip-base.js [hunk 15]
```diff
@@ -0,0 +1,171 @@
+import $ from "jquery";
+import Backbone from "backbone";
+Backbone.$ = $;
+
+import "../../css/ui/tooltips.css";
```
Comment (kevinbarabash) [inline js/ui/tooltip-base.js]: I see, you're importing the .css directly in the .js files.  Does this result in the .css appearing in the resulting bundles?

File: js/ui/tooltip-base.js (hunk 2 above)
Comment (kevinbarabash) [reply js/ui/tooltip-base.js]: Cool.  I'm going to assume that there are no changes `TooltipBase`.

File: js/ui/tooltip-engine.js [hunk 16]
```diff
@@ -188,166 +195,8 @@ window.TooltipEngine = Backbone.View.extend({
...
-        this.$el
-            .css({
-                top: $(window).scrollTop() + coords.pageY,
-                left: coords.pageX
-            })
-            .toggle(!(relativePos < 0 || relativePos >= editorHeight));
-    },
+TooltipEngine.registerTooltip = function(name, tooltipClass) {
+    TooltipEngine.classes[name] = tooltipClass;
+};
```
Comment (kevinbarabash) [inline js/ui/tooltip-engine.js]: You could also do:
```
export registerTooltip = (name, tooltipClass) => TooltipEngine.classes[name] = tooltipClass;
```
and then in files that use `registerTooltip` do:
```
import {registerTooltip} from "./tooltip-engine.js";
```
this is more idiomatic.

File: tests/tooltips/colorPicker_test.js (hunk 7 above)
Comment (kevinbarabash) [reply tests/tooltips/colorPicker_test.js]: How does sinon's `sandbox` differ from what we were doing before?

Comment (kevinbarabash) [reply tests/tooltips/colorPicker_test.js]: Also, these sandbox changes probably would made a good PR on their own.

File: tmpl/live-editor.handlebars (hunk 8 above)
Comment (kevinbarabash) [reply tmpl/live-editor.handlebars]: Does this mean we can get rid of the handlebars plugin which was handling `{{#_}}`/`{{/_}}`?

File: webpack.config.js [hunk 17]
```diff
@@ -65,9 +91,33 @@ const config = {
...
+                }
+            },
+            // Needed for loading the Bootstrap font file with WOFF/SVGs
+            {
+                test: /\.(woff|woff2|eot|ttf|svg)$/,
+                loader: 'url-loader?limit=100000'
+            },
         ],
     },
     plugins: [
```
Comment (kevinbarabash) [inline webpack.config.js]: Usually people extract the CSS back out of the bundles using either ExtractTextPlugin or MiniCssPlugin.  I'm cool with not doing that.  I'm not sure how keeping the CSS in the bundles affects our ability to override it in webapp.  Maybe that's necessary give that we're using different templates in webapp.

File: js/ui/tooltips/auto-suggest.js (hunk 3 above)
Comment (pamelafox) [reply js/ui/tooltips/auto-suggest.js]: Good Q. Note that we still pop-up the function arguments hint, but we disabled the auto completion of functions. (Both are handled by ScratchpadAutoSuggest currently)

See comment on https://github.com/Khan/live-editor/blob/a367b6383f9cb1e984935939635e1a80cea5d445/js/ui/tooltip-engine.js#L113 
Here's the commit where I disabled it:
https://github.com/Khan/live-editor/commit/4e028ca5d15d3b062a4a9b38fd4033ddb1a75758

Autosuggest was a hack project, as I recall, so it didn't have time to get well integrated and tested. It'd be better if it could have more time spent on it.

File: tests/tooltips/colorPicker_test.js (hunk 7 above)
Comment (pamelafox) [reply tests/tooltips/colorPicker_test.js]: Good call, these changes can go straight to master, I've sent a PR.

**Comment by pamelafox** [comment]
Closing since I'm no longer at KA to make sure this would work within webapp too.
